src/base64.cr

```diff
@@ -245,74 +394,202 @@ module Base64
+
+      chunk : UInt32 = 0
+      bytes_state : UInt8 = 0
+
+      # Read and decode characters into `chunk`,
+      # writing additional flags to `bytes_state`.
+      current_byte = decode_ptr[in_ptr.value]
+      chunk = (chunk << 6) &+ current_byte
+      bytes_state |= current_byte
+      in_ptr += 1
+      current_byte = decode_ptr[in_ptr.value]
+      chunk = (chunk << 6) &+ current_byte
+      bytes_state |= current_byte
+      in_ptr += 1
+      current_byte = decode_ptr[in_ptr.value]
+      chunk = (chunk << 6) &+ current_byte
+      bytes_state |= current_byte
+      in_ptr += 1
+      current_byte = decode_ptr[in_ptr.value]
+      chunk = (chunk << 6) &+ current_byte
+      bytes_state |= current_byte
+      in_ptr += 1
+
+      raise Base64::Error.new("Invalid base64 chunk") if bytes_state >= 0x80_u8
+      if bytes_state >= 0x40_u8
+        in_ptr -= 4
+        break
       end
-      break if bytes > fin
 
-      yield_decoded_chunk_bytes(bytes[0], bytes[1], bytes[2], bytes[3], chunk_pos: bytes - bytes_begin)
-      bytes += 4
+      # Write resulting bytes
+      out_ptr.value = (chunk >> 16).to_u8!
+      out_ptr += 1
+      out_ptr.value = (chunk >> 8).to_u8!
+      out_ptr += 1
+      out_ptr.value = chunk.to_u8!
+      out_ptr += 1
     end
 
-    # Move the pointer by one byte until there is a valid base64 character or the end of `bytes` was reached
-    while (bytes < fin + 4) && bytes.value.in?(NL, NR)
-      bytes += 1
+    {Int32.new!(in_ptr - in_ptr_orig), Int32.new!(out_ptr - out_ptr_orig)}
+  end
+
+  # Decodes base64 data from the `in` buffer and writes it into the `out` buffer.
+  #
+  # This method expects the buffer given via `in_ptr` and `in_size`
+  # to contain exactly one encoding character (with optional padding at the end).
+  #
+  # Once this method has written at least one byte into the output buffer,
+  # it may not be called again for the remaining part of the buffer.
+  # For this, you can use `consume_buffer_garbage`.
+  # That being said, note that there can only be a "remaining part of the buffer" if
+  # not all relevant bytes have been given to `decode_buffer_final`.
+  #
+  # Raises a `Base64::Error` when invalid chunks/characters are passed.
+  #
+  # Returns the amount of bytes read and the amount of bytes written.
+  #
+  # NOTE: This method expects the buffer used for decoding to contain
+  #       all decodable characters if there is no padding before it.
+  #       If the given buffer contains only a part of the last chunk without padding,
+  #       either an error is thrown or the output is cut off.
+  private def decode_buffer_final(in_ptr : UInt8*, in_size : Int32, out_ptr : UInt8*, out_size : Int32) : Tuple(Int32, Int32)
+    in_pos : Int32 = 0
+    decode_ptr : UInt8* = DECODE_TABLE.to_unsafe
+
+    # Move the pointer by one byte until there is a valid base64 character
+    while in_pos < in_size && in_ptr.value.in?(10_u8, 13_u8)
+      in_pos &+= 1
```

Comment:
This seems odd. Contrary to the comment, this loop does not move the pointer. It just increases `in_pos` which is simple integer counter. Should this include `in_ptr += 1` perhaps?